### ISSUE TITLE
DBUtils' problem with pymssql 

### ISSUE THREAD

**Issue by 421913533yw** [WebwareForPython/DBUtils#52]
DBUtils has a problem, I use pymssql as the database engine, sql server database has a table field length of 2, and there is a record, when I go to update this record, and the length of the write is more than 2, The entire process will be blocked, causing the interface card to die (similarly, any update that exceeds the field length will do so and cannot be stopped unless time_out is set, which is not what I expect). The version of DBUtils I use is 1.3 and 3.1 both have the same problem.
After debugging, the code blocked at [result = method2(*args, **kwargs)], and I tried to remove lines 633-678 of the steady_db.py file, which resolved the problem.

**Comment by Cito** [comment]
Can you post reproducible example code?

**Comment by 421913533yw** [comment]
I use pymssql as the connection client in the code. There is a table in the sqlserver database called TEST_INSERT_TABLE, which has the fields ID (integer) and BIG_DATA (VARCHAR of length 2). First, I can successfully insert A record (1, "A"), then update the record with ID 1 to change BIG_DATA to ABCDEFG, and when I execute this command, the program will block, which is when I first raised the issue.
Here is the sample code
```
import pymssql
from dbutils.pooled_db import PooledDB

if __name__ == "__main__":
    pool = PooledDB(creator=pymssql,
                    mincached=3,
                    maxcached=10,
                    maxshared=5,
                    maxconnections=12,
                    blocking=True,
                    ping=1,
                    charset="utf8",
                    host="127.0.0.1",
                    port=1433,
                    database="test_db",
                    user="sa",
                    password="123456")
    conn = pool.connection()
    cursor = conn.cursor()
    cursor.execute("INSERT INTO TEST_INSERT_TABLE(ID, BIG_DATA) VALUES(%d, %s)", (1, "A"))
    cursor.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", (1, "ABCDEFG"))
    conn.commit()
    conn.close()
```

**Comment by Cito** [comment]
Sorry, I cannot reproduce this.

Your script gives the error `Conversion failed when converting the nvarchar value 'ABCDEFG' to data type int`. This is because you have the wrong order of elements in the updated tuple. If I fix the order, then I get the error `String or binary data would be truncated in table 'TEST_INSERT_TABLE', column 'BIG_DATA'`. Which is expected because the column allows only 2 chars but you try to store 7.

Tested with:
- Python 3.12
- DBUtils 3.1.0
- PyMSSQL 2.3.1
- SQL Server 16.0.1000.6 (Express version)

**Comment by 421913533yw** [comment]
I'm sorry for making this code error. As you described, when you debug the code, the first exception thrown is indeed caught and a second attempt is made, but what hangs on the second attempt is line 654 of steady_db.py, to be precise, the code 'result = method2(*args, **kwargs)'.

This is the corrected code：
```
import pymssql
from dbutils.pooled_db import PooledDB

if __name__ == "__main__":
    pool = PooledDB(creator=pymssql,
                    mincached=3,
                    maxcached=10,
                    maxshared=5,
                    maxconnections=12,
                    blocking=True,
                    ping=1,
                    charset="utf8",
                    host="127.0.0.1",
                    port=1433,
                    database="test_db",
                    user="sa",
                    password="123456")
    conn = pool.connection()
    cursor = conn.cursor()
    cursor.execute("INSERT INTO TEST_INSERT_TABLE(ID, BIG_DATA) VALUES(%d, %s)", (1, "A"))
    conn.commit()

    cursor.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
    conn.commit()
    conn.close()
```
This is my database table field type and the associated configuration
![1730445649222](https://github.com/user-attachments/assets/0da11a11-9e63-4904-8401-d9faf83d3862)


Tested with:
  - Python 3.9.13
  - DBUtils 3.1.0
  - pymssql 2.2.7
  - database：SQL Server 2022 - 16.0.1000.6 (X64)  and SQL Server 2019 - 15.0.2000.5 (X64) 
 
Your framework is very useful, but this question has puzzled me for a long time, good luck!
I would greatly appreciate it if you could provide more assistance!

**Comment by Cito** [comment]
Thanks for the feedback. I have tried with the exact same versions (Python 3.9 and PyMSSQL 2.2.7), and set up the table exactly as yours  and still cannot reproduce the problem. I was not able to select the exact same Chinese collation and used a different Chinese one, but I don't think that can make a difference.

When I run your example, the first exception thrown is indeed caught and a second attempt is made, but it does not hang, instead it produces the expected error message "String or binary data would be truncated in table 'test.dbo.TEST_INSERT_TABLE', column 'BIG_DATA'."

Can you think of anything else that could be different or special in your setup? Maybe the operation system? I have tested this under Windows 11.

**Comment by 421913533yw** [comment]
Well, there's one key point I forgot to mention, did you set the ID column as the primary key? When the ID column is not the primary key, the code can throw an exception normally, and when the ID column is an exception, the code will hang.
![1730690667689](https://github.com/user-attachments/assets/3bc37a41-c7f5-4110-b7b8-03a817599f73)

I have also tested the platform issues you mentioned, I have tested Windows 10, Windows11, Ubuntu 22.04.2 LTS, Docker containers have such problems.

**Comment by 421913533yw** [comment]
I tried using only pymssql, tested it to mimic the form of your framework (based on a table with ID as the primary key), and constructed three code blocks (blocks 1, 2, and 3). The construction logic is:

Code block 1: First initialize the first connection (conn1), when the first update throws an exception, do not close the current connection, initialize a connection (conn2), perform the update operation, then the second connection is suspended (very similar to the phenomenon with dbutils);

**blocks 1**
```
import pymssql

if __name__ == "__main__":
    conn1 = None
    try:
        conn1 = pymssql.connect(server="127.0.0.1", user="sa", password="q1w2e3r4t5", database="DBZS", port=1433)
        cursor1 = conn1.cursor(as_dict=True)
        cursor1.execute("INSERT INTO TEST_INSERT_TABLE(ID, BIG_DATA) VALUES(%d, %s)", (1, "A"))
        conn1.commit()
        cursor1.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
        conn1.commit()
    except Exception as e:
        print("frist-error:", e)
        # if conn1:
        #     conn1.close()
        try:
            conn2 = pymssql.connect(server="127.0.0.1", user="sa", password="q1w2e3r4t5", database="test_db", port=1433)
            cursor2 = conn2.cursor(as_dict=True)
            cursor2.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
            conn2.commit()
        except Exception as ee:
            print("second-error:", ee)
```


Code block 2: First initialize the first connection (conn1), when the first update throws an exception, close the conn1 connection, and then initialize a connection (conn2), perform the update operation, then the exception of the second connection is successfully thrown;

**blocks 2**
```
import pymssql

if __name__ == "__main__":
    conn1 = None
    try:
        conn1 = pymssql.connect(server="127.0.0.1", user="sa", password="q1w2e3r4t5", database="test_db", port=1433)
        cursor1 = conn1.cursor(as_dict=True)
        cursor1.execute("INSERT INTO TEST_INSERT_TABLE(ID, BIG_DATA) VALUES(%d, %s)", (1, "A"))
        conn1.commit()
        cursor1.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
        conn1.commit()
    except Exception as e:
        print("frist-error:", e)
        if conn1:
            conn1.close()
        try:
            conn2 = pymssql.connect(server="127.0.0.1", user="sa", password="q1w2e3r4t5", database="test_db", port=1433)
            cursor2 = conn2.cursor(as_dict=True)
            cursor2.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
            conn2.commit()
        except Exception as ee:
            print("second-error:", ee)
```


Code block 3: First initialize the first connection (conn1), when the first update throws an exception, do not close the conn1 connection, perform an update operation again, then the second exception is successfully thrown;

**blocks 3**
```
import pymssql

if __name__ == "__main__":
    conn1 = None
    try:
        conn1 = pymssql.connect(server="127.0.0.1", user="sa", password="q1w2e3r4t5", database="DBZS", port=1433)
        cursor1 = conn1.cursor(as_dict=True)
        cursor1.execute("INSERT INTO TEST_INSERT_TABLE(ID, BIG_DATA) VALUES(%d, %s)", (1, "A"))
        conn1.commit()
        cursor1.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
        conn1.commit()
    except Exception as e:
        print("frist-error:", e)
        try:
            cursor1.execute("UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
            conn1.commit()
        except Exception as ee:
            print("second-error:", e)
```

In summary, I guess whether it is because the first connection is not closed, and the second connection is opened to perform the update operation, resulting in the suspension. I suspect that this may be the result of no post-exception handling on line 613 of steady_db.py.
![1730701188634](https://github.com/user-attachments/assets/529757e4-96f3-4b0d-83b3-7542230dea7d)

**Comment by Cito** [comment]
Thanks for the feedback. After making the ID column the primary key I can reproduce the problem now, both with the older and newer versions. Will try to look into it this week.

**Comment by Cito** [comment]
Looked into this again today and found the cause of the problem: You need to explicitly call `begin()` at the start of every transaction to avoid such problems. This is a peculiarity of DBUtils, as explained in the documentation:

> Note that you need to explicitly start transactions by calling the begin() method. This ensures that the connection will not be shared with other threads, that the transparent reopening will be suspended until the end of the transaction, and that the connection will be rolled back before being given back to the connection pool.

I have made this paragraph stand out more clearly in the [documentation](https://webwareforpython.github.io/DBUtils/main.html) now - you may need to reload the page to get the new version with the new CSS.

So this should work:

```python
conn = pool.connection()
cursor = conn.cursor()

conn.begin()
cursor.execute(
    "DELETE FROM TEST_INSERT_TABLE")
cursor.execute(
    "INSERT INTO TEST_INSERT_TABLE(ID, BIG_DATA) VALUES(%d, %s)", (1, "A"))
conn.commit()

conn.begin()
cursor.execute(
    "UPDATE TEST_INSERT_TABLE SET BIG_DATA=%s WHERE ID=%d", ("ABCDEFG", 1))
conn.commit()
```

Please let me know whether this solves your problem.

**Comment by 421913533yw** [comment]
I tried the method you described and it really solved my problem, thank you very much! But when I try to do the same thing with pymssql alone (see code blocks 1), the program hangs the same, which I can't understand because the package doesn't have a begin method.

**Comment by Cito** [comment]
In your code block1 you're doing exactly what DBUtils did when there was no `begin()`. Seems like SQL Server waits until the (implicit) transaction is committed or rolled back before doing the update, but since that never happens, the program hangs. Note that the `begin()` does not do anything on the server, it just tells DBUtils that you started to use the connection in a transaction.

**Comment by 421913533yw** [comment]
Thank you very much for your patience and helpful guidance, which has greatly benefited me. Good luck to you!
